src/rebar_packages.erl

```diff
@@ -22,57 +25,93 @@
+    case rebar_prv_update:do(Registry, State) of
+        {ok, State1} ->
+            case load_and_verify_version(Registry, State1) of
+                {ok, T} ->
+                    {ok, T};
+                false ->
+                    %% Still unable to load after an update, create an empty registry
+                    ets:new(repo, [public])
+            end;
+        E ->
+            throw(E)
     end.
 
-close_packages() ->
-    catch ets:delete(?PACKAGE_TABLE).
+close_packages(T) ->
+    catch ets:delete(T).
 
-load_and_verify_version(State) ->
-    {ok, RegistryDir} = registry_dir(State),
-    case ets:file2tab(filename:join(RegistryDir, ?INDEX_FILE)) of
-        {ok, _} ->
-            case ets:lookup_element(?PACKAGE_TABLE, package_index_version, 2) of
-                ?PACKAGE_INDEX_VERSION ->
-                    true;
-                _ ->
-                    (catch ets:delete(?PACKAGE_TABLE)),
-                    rebar_prv_update:hex_to_index(State)
+load_and_verify_version(Registry, State) ->
+    case ?MODULE:registry_dir(Registry, State) of
+        {ok, RegistryDir} ->
+            case ets:file2tab(filename:join(RegistryDir, ?INDEX_FILE)) of
+                {ok, T} ->
+                    {ok, T};
+                E ->
+                    E
             end;
-        _ ->
-            rebar_prv_update:hex_to_index(State)
+        E ->
+            E
     end.
 
 deps(Name, Vsn, State) ->
     try
-        deps_(Name, Vsn, State)
+        find(rebar_state:repos(State),
+            fun(Registry) -> deps_(Name, Vsn, Registry, State) end)
     catch
         _:_ ->
             handle_missing_package({Name, Vsn}, State, fun(State1) -> deps_(Name, Vsn, State1) end)
     end.
 
 deps_(Name, Vsn, State) ->
-    ?MODULE:verify_table(State),
-    ets:lookup_element(?PACKAGE_TABLE, {ec_cnv:to_binary(Name), ec_cnv:to_binary(Vsn)}, 2).
+    find(rebar_state:repos(State),
+         fun(Registry) -> deps_(Name, Vsn, Registry, State) end).
+
+deps_(Name, Vsn, Registry, State) ->
+    {ok, T} = ?MODULE:verify_table(Registry, State),
+    case ets:lookup_element(T, {ec_cnv:to_binary(Name), ec_cnv:to_binary(Vsn)}, 2) of
+        [Deps | _] ->
+            {ok, Deps};
+        E ->
+            E
+    end.
+
+find_package_registry(Name, Vsn, State) ->
+    find(rebar_state:repos(State),
+         fun(Registry) -> find_package_registry_(Name, Vsn, Registry, State) end).
+
+find_package_registry_(Name, Vsn, Registry, State) ->
+    {ok, T} = ?MODULE:verify_table(Registry, State),
+    case ets:member(T, {ec_cnv:to_binary(Name), ec_cnv:to_binary(Vsn)}) of
+        true ->
+            {ok, Registry};
+        false ->
+            not_found
+    end.
+
+find([], _) ->
+    error;
+find([X | R], Fun) ->
+    case catch(Fun(X)) of
+        {ok, Y, Z} ->
+            {ok, Y, Z};
+        {ok, Y} ->
+            {ok, Y};
+        _ ->
+            find(R, Fun)
+    end.
```

Comment:
replace `case catch(X) of` by `try X of` for faster processing if we ignore the failing cases. Will be faster, especially in projects with many layers of repos.

Not too sure what the failing case would be there? Just wanted a single clause for exceptions and a package not being found?